File: src/leadforge/scrapers/craigslist.py
import structlog

from leadforge.scrapers.base import BaseAPIClient

logger = structlog.get_logger()


class CraigslistClient(BaseAPIClient):
    """Craigslist Chicago services scraper."""

    def __init__(self):
        super().__init__(base_url="https://chicago.craigslist.org", timeout=30.0)

    async def search_services(self, name: str, category: str = "bbb") -> dict | None:
        """Search Craigslist services for a business. Returns listing data or None."""
        try:
            client = await self._get_client()
            response = await client.get(
                f"/search/{category}",
                params={"query": name},
            )
            response.raise_for_status()
            text = response.text

            # Count listings mentioning the business
            import re

            listings = re.findall(r'class="result-title[^"]*"[^>]*>([^<]+)', text)
            matching = [item for item in listings if name.lower() in item.lower()]

            return {
                "craigslist_listing_count": len(matching),
                "craigslist_has_presence": len(matching) > 0,
            }
        except Exception as e:
            logger.warning("craigslist_search_failed", error=str(e), name=name)
            return None

File: api/src/scrapers/craigslist.ts
// =py scrapers/craigslist
import { encodeParams, fetchText } from './base';

const RESULT_TITLE_RE = /class="result-title[^"]*"[^>]*>([^<]+)/g;

export interface CraigslistPresence {
  craigslist_listing_count: number;
  craigslist_has_presence: boolean;
}

// =py search_services — informational only; the enrichment caller logs the result and stores nothing
export async function searchServices(name: string, category = 'bbb'): Promise<CraigslistPresence | null> {
  try {
    const html = await fetchText(`https://chicago.craigslist.org/search/${category}?${encodeParams({ query: name })}`);

    const titles = [...html.matchAll(RESULT_TITLE_RE)].map((m) => m[1]);
    const matching = titles.filter((title) => title.toLowerCase().includes(name.toLowerCase()));

    return {
      craigslist_listing_count: matching.length,
      craigslist_has_presence: matching.length > 0,
    };
  } catch (error) {
    console.warn('craigslist_search_failed', { name, error: error instanceof Error ? error.message : String(error) });
    return null;
  }
}
